---
title: "BERA_20170817_RNA"
author: "Ben K Johnson"
date: "September 19, 2017"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


```{r loaddata}

library(tximport)
library(readr)
library(tidyr)
library(plyr)
library(wasabi)


#Define the base directory location; similar to setwd()
base_dir <- "~/secondary_projects/research/BERA_20170817_RNA/analysis/salmon"

#Get the sample ids
sample_id <- list.files(base_dir, pattern = "*_quant") %>% {gsub("\\_.*", "", .)}
sample_id <- as.character(sample_id)

#Get the file paths for each quant file in salmon output
file_paths <- sapply(sample_id, function(id) file.path(base_dir, paste0(id, "_quant"), "quant.sf"))

#Clean the data by removing the isoform information
#WARNING: THIS OVERWRITES THE quant.sf FILE!!!
cleaned_data <- lapply(file_paths, function(x) {
  indat <- read.delim(x, header = T)
  indat$Name <- gsub("\\..*", "", indat$Name)
  write.table(indat, x, sep = '\t', quote = F, row.names = F)
})

############Below is the way to convert Salmon output for use with Sleuth####################
##NOTE: YOU WILL WANT TO USE BOOTSTRAPS OVER GIBBS HERE. IT'S UNCLEAR WHETHER GIBBS ARE IMPORTED BY SLEUTH APPROPRIATELY.

#Convert the salmon input to sleuth compatible input
#Only need to run this once to convert to abundance.h5 files
prepare_fish_for_sleuth(dir(base_dir, pattern = "*_quant", include.dirs = T, full.names = T))

#Set up conditions
dir_paths <- sapply(sample_id, function(id) file.path(base_dir, paste0(id, "_quant")))
condition <- as.data.frame(cbind(sample_id, c(rep("S", 6), rep("X", 8)), dir_paths), stringsAsFactors = FALSE)
colnames(condition) <- c("sample", "condition", "path")
row.names(condition) <- seq(1:14)

#############################################################################################


###########Import data with STAR 2-pass reverse gene counts############################

#Read in the HTSeq combined matrix
htseq.raw <- read.delim("/Volumes/projects_secondary/bbc/research/BERA_20170817_RNA/analysis/star/merged_reads/joinedSTAR2pass.withheader.count.matrix", stringsAsFactors = FALSE, check.names = FALSE)

#Set as rownames
rownames(htseq.raw) <- htseq.raw$ens_gene

#Remove the ens_gene column
htseq.raw$ens_gene <- NULL

#Format colnames
colnames(htseq.raw) <- gsub("\\ReadsPerGene.out.tab", "", colnames(htseq.raw))

#Remove the last column since it contains NAs
htseq.raw <- htseq.raw[,-17]

```

```{r, dte}

#############Below is the way to do DTE with Sleuth##########################################
library(sleuth)
library(cowplot)

#Define genotype information
condition$condition <- factor(condition$condition)
condition$condition <- relevel(condition$condition, ref = "S")


#Get the annotations
library(biomaRt)
mart <- biomaRt::useMart(biomart = "ENSEMBL_MART_ENSEMBL", dataset = "rnorvegicus_gene_ensembl", host="useast.ensembl.org")
t2g <- biomaRt::getBM(attributes = c("ensembl_transcript_id", "ensembl_gene_id", "mgi_symbol", "description", "entrezgene"), mart = mart)
t2g <- dplyr::rename(t2g, target_id = ensembl_transcript_id, ens_gene = ensembl_gene_id, ext_gene = mgi_symbol, predicted_function = description, entrez_id = entrezgene)

#Sleuth prep

full_design <- model.matrix(~condition$condition)
colnames(full_design) <- c("Intercept", "Genotype")

#My basic filter to apply to sleuth prep
#50% of samples must have at least 10 est counts
my_basic_filter <- function(row, min_reads = 10, min_prop = 0.5) {
  mean(row >= min_reads) >= min_prop
}

so <- sleuth_prep(condition, full_model = full_design, target_mapping = t2g, filter_fun = my_basic_filter) %>% sleuth_fit()

#Fit the reduced model
so <- sleuth_fit(so, formula = ~1, fit_name = "reduced")

#View the models
models(so)

#LRT
so <- sleuth_lrt(so, "reduced", "full")

sleuth_table <- sleuth_results(so, 'reduced:full', 'lrt', show_all = FALSE)
sleuth_significant <- dplyr::filter(sleuth_table, qval <= 0.05)

write.table(sleuth_significant, "~/secondary_projects/research/BERA_20170817_RNA/analysis/sleuth/sleuth_lrt_q005.txt",
            sep = '\t', quote = F, row.names = F)

#Wald test
so <- sleuth_wt(so, which_model = "full", which_beta = "Genotype")

sleuth_table_wt <- sleuth_results(so, 'Genotype', 'full', show_all = FALSE)
sleuth_significant_wt <- dplyr::filter(sleuth_table_wt, qval <= 0.05)

write.table(sleuth_significant_wt, "~/secondary_projects/research/BERA_20170817_RNA/analysis/sleuth/sleuth_wt_q005.txt",
            sep = '\t', quote = F, row.names = F)

#go to sleuth live
sleuth_live(so)
#############################################################################################

```


```{r lib-size}

##The following is using the STAR HTSeq matrix as input for traditional DGE

library(limma)
library(edgeR)
library(ggplot2)
library(plyr)
library(dplyr)
library(reshape2)
library(RColorBrewer)
library(ggrepel)
library(grid)

# Calculate library sizes & transform for plotting
lib_sizes <- data.frame(colSums(htseq.raw)/1000000)
lib_sizes$sample <- factor(rownames(lib_sizes), levels = rownames(lib_sizes))
colnames(lib_sizes) <- c("Size", "Sample")

ggplot(lib_sizes, aes(x = Sample, y = Size)) +
  geom_bar(stat="identity") +
  ylab("Raw Counts (Millions)") +
  #scale_y_continuous(limits = c(0, 25)) +
  scale_fill_brewer(palette = "Paired") +
  ggtitle("Raw Counts") +
  theme_bw(10) +
  theme(
    axis.text.x = element_text(angle=45, hjust=1),
    axis.title.x = element_blank(),
    plot.title = element_text(hjust = 0.5)
  )

```

```{r distributions, fig.height=10, fig.width=18}

#Filter by CPM >= 2
x <- htseq.raw[rowSums(cpm(as.matrix(htseq.raw)) > 1) >= 8, ]


# Voom transform
treatment <- factor(c(rep("Saline", 8), rep("MDMA", 8)))
treatment <- relevel(treatment, ref = "Saline")

#Fit the full model - written explicitly to make it easier to read
design <- model.matrix(~treatment)
rownames(design) <- rownames(lib_sizes)

#Convert to a DGEList object to use TMM normalization in voom transform
x <- DGEList(counts = x, group = treatment)

#Adjust for library size differences
x <- calcNormFactors(x)

y <- voom(x, design)

v2 <- t(y$E)

stopifnot(all(colnames(htseq.raw)==rownames(v2)))

##Plot the genotype specific count distributions adjusting for toc and sex
v2 <- cbind(lib_sizes, v2)
v2$Size <- NULL
#v2$Sample <- NULL
v2$Treatment <- treatment
melted <- melt(v2, id.var=c("Sample", "Treatment"))
colnames(melted) <- c("Sample", "Treatment", "Gene", "value")

# Plot transformed count distributions
ggplot(melted, aes(x=Sample, y=value, fill=Treatment)) +
  geom_boxplot() +
  scale_fill_brewer(palette = "Set1") +
  ylab("log2 Counts Per Million") +
  ggtitle("Transformed Count Distributions") +
  theme_bw(10) +
  theme(
    axis.text.x = element_text(angle=45, hjust=1),
    axis.title.x = element_blank(),
    plot.title = element_text(hjust = 0.5)
  )

```

## PCA 

```{r pca, fig.height=10, fig.width=17}

##Plot the PCA of genotype specific effects adjusting for toc and sex
pca <- prcomp(t(y$E))
pr_comps <- data.frame(pca$x)

stopifnot(all(colnames(htseq.raw) == rownames(pr_comps)))

# Combine for plotting
pr_comps$Sample <- rownames(pr_comps)

pr_comps$Treatment <- factor(v2$Treatment)

cols <- colorRampPalette(brewer.pal(6, "Set1"))

pca_plot <- ggplot(pr_comps, aes(x=PC1, y=PC2, color=Treatment)) + 
  geom_point(size=3.5) + 
  ylim(-100, 100) +
  xlim(-100, 100) +
  geom_text_repel(aes(label = as.character(Sample))) +
  scale_fill_manual(values=myPal) + 
  theme_bw(10)

# Plot percent variation explained
prop_var <- data.frame(t(summary(pca)$importance))
names(prop_var) = c('sd', 'prop', 'cum')
prop_var$num = 1:nrow(prop_var)

var_plot <- ggplot(prop_var, aes(x=num, y=prop)) + 
  geom_point(size=3.5) + 
  geom_line() + 
  scale_x_continuous(limits = c(1, 16), breaks = 1:16) +
  xlab("Principal Component") + 
  ylab("Prop. of Variance") +
  theme_bw(10) +
  theme(
    axis.title.y = element_text(vjust=1),
    plot.margin = unit(c(0,0,0,6), "mm")
  )

vplayout <- function(x, y) viewport(layout.pos.row = x, layout.pos.col = y)

grid.newpage()
pushViewport(viewport(layout = grid.layout(4, 100)))
print(pca_plot, vp = vplayout(1:3, 3:100))
print(var_plot, vp = vplayout(4, 1:87))

```

```{r plotNfit}

#Plot the MDS plot for genotype specific differences adjusting for toc and sex
my.mdsobj <- plotMDS(y, top=500, gene.selection="pairwise", main="All samples", plot = FALSE)

my.mdsobj.cmdscale <- as.data.frame(my.mdsobj$cmdscale.out)

my.mdsobj.cmdscale$Treatment <- treatment

cols <- colorRampPalette(brewer.pal(7, "Paired"))

ggplot(my.mdsobj.cmdscale, aes(x=V1, y=V2, color=Treatment, shape=Treatment)) + 
  geom_point(size=10) + 
  ylim(-4, 4) +
  xlim(-4, 4) +
  ylab(bquote('Leading' ~Log[2]~ 'Fold Change Dim 2')) +
  xlab(bquote('Leading' ~Log[2]~ 'Fold Change Dim 1')) +
  scale_colour_brewer(type="qual", palette="Paired") +
  #ggtitle("MDS Plot of Expression Profiling") +
  theme_bw(10) +
  theme(
    #plot.title = element_text(hjust = 0.5),
    text = element_text(size = 22)
)

```

```{r edgeR}

##Generate the DGEList object
y <- DGEList(counts = htseq.raw, group = treatment)

#Filter the data, keeping anything that has a CPM > 2 in at least half the samples
keep <- rowSums(cpm(y)>1) >= 8

#Apply the filter
y <- y[keep, , keep.lib.sizes = FALSE]

#Normalize based on library size
y <- calcNormFactors(y)

#Calc dispersions
y <- estimateDisp(y, design, robust = TRUE)

#Check the dispersion estimates
y$common.dispersion

#Plot the BCV dispersions
plotBCV(y)

#Fit using the quasi-liklihood functionality
fit <- glmQLFit(y, design, robust=TRUE)

#Plot the quasi-liklihood dispersions
plotQLDisp(fit)
```

```{r dge}

#Test for differential expression given genotype
qlf <- glmQLFTest(fit)

#Look at the top dge
topTags(qlf)

#Grab the dge table and only keep dge q < 0.05
tt <- topTags(qlf, n = Inf, p.value = 0.05, sort.by = "PValue")

#Only keep the table
tt <- tt$table

#Generate the ens_gene column to annotate
tt$ens_gene <- rownames(tt)

#How many genes are dge with q < 0.05
summary(dt <- decideTestsDGE(qlf))

#Plot the smear plot
isDE <- as.logical(dt)
DEnames <- rownames(y)[isDE]

plotSmear(qlf, de.tags=DEnames)
#Add in lines for 2-fold differences
abline(h=c(-1,1), col="blue")
```

```{r, annotate}

# Annotations for data tables

tt.annot <- plyr::join(tt, t2g, by = "ens_gene") 

tt.annot <- subset(tt.annot, !duplicated(tt.annot$ens_gene))

write.table(tt.annot, "/Volumes/projects_secondary/bbc/research/BERA_20170817_RNA/analysis/edgeR/BERA_20170817_q005_dge.txt", sep = '\t', quote = F, row.names = F)
```

```{r, largestfc}

##Extract the genes with at least a 2-fold change
tt.annot.largefc <- subset(tt.annot, tt.annot$logFC >= 1 | tt.annot$logFC <= -1)

##Write out the table
#write.table(tt.annot.largefc, "/Volumes/projects_secondary/bbc/research/PFEG_20170623_RNA/analysis/PFEG20170623_greaterthan_2folddiff_dge.txt", sep = '\t', quote = F, row.names = F)

##Import the GO term data from the 2-fold change genes and plot
library(cowplot)
library(fgsea)

rat.gmt <- "/Volumes/projects_secondary/bbc/research/BERA_20170817_RNA/analysis/edgeR/Rattus_norvegicus_GSEA_GO_sets_all_ids_April_2015.gmt"

pathways <- gmtPathways(rat.gmt)

saline_v_mdma <- tt.annot

saline_v_mdma <- dplyr::mutate(saline_v_mdma, fcSign = sign(logFC),
                               logP = -log10(PValue),
                               t = logP / fcSign,
                               ID = entrez_id)

saline_v_mdma_rnk <- saline_v_mdma[,c("ID", "t")]

saline_v_mdma_rnk$ID <- as.character(saline_v_mdma_rnk$ID)

saline_v_mdma_rnk <- setNames(saline_v_mdma_rnk$t, saline_v_mdma_rnk$ID)

gsea <- fgsea(pathways, saline_v_mdma_rnk, minSize = 5,
              maxSize = 500, nperm = 1000)

#THERE IS 0 GO ENRICHMENT

```

```{r heatmaps}

##Generate a heatmap of the top DGE with a minimum of 2-fold differences
cd <- htseq.raw[rowSums(cpm(as.matrix(htseq.raw)) > 2) >= 3, ]
v <- voom(cd, design)
exprs <- as.data.frame(v$E)
ratio <- sweep(exprs, 1, apply(exprs[, grep("WT*", colnames(exprs))], 1, mean, na.rm = TRUE), "-")

row.names(ratio) <- gsub("\\..*", "", rownames(ratio))

heatde <- ratio[tt.annot.largefc$ens_gene,]
heatde$ens_gene <- rownames(heatde)
heatde <- plyr::join(heatde, t2g, by="ens_gene")
heatde <- subset(heatde, !duplicated(heatde$ens_gene))

brk <- seq(-3,3,0.06)

ind_lfc.m <- as.matrix(heatde[,1:6])
ind_lfc.m[ind_lfc.m > 3] <- 3
ind_lfc.m[ind_lfc.m < -3] <- -3
rownames(ind_lfc.m) <- heatde$ext_gene

library(pheatmap)

#Change pheatmap's way of adding labels to angle at 45
draw_colnames_45 <- function (coln, ...) {
    m = length(coln)
    x = (1:m)/m - 1/2/m
    grid.text(coln, x = x, y = unit(0.96, "npc"), vjust = .75, 
        hjust = 1, rot = 45, gp = gpar(...)) ## Was 'hjust=0' and 'rot=270'
}

assignInNamespace(x="draw_colnames", value="draw_colnames_45", ns=asNamespace("pheatmap"))

Genotype <- c("darkgrey", "lightgrey")
names(Genotype) <- c("WT", "Smchd1KO")
ann_colors = list(Genotype=Genotype)

heatmap_geno_anno <- as.matrix(c("WT", "WT", "WT", "Smchd1KO", "Smchd1KO", "Smchd1KO"))
colnames(heatmap_geno_anno) <- "Genotype"
merged_types <- as.data.frame(cbind(heatmap_geno_anno))
rownames(merged_types) <- colnames(ind_lfc.m)

# Define colors
my_palette <- colorRampPalette( c("blue","white","red"))(100)

#tiff(file = "~/Documents/VANJ_20161220_RNA/deliverables/WTvsmitomutants_DEgenes_rowclust_heatmap_p05_withanno.tiff", width=8, height=12, units='in', res=600)
pheatmap(ind_lfc.m, border_color=NA, cluster_rows=TRUE, cluster_cols=TRUE, Rowv=NA, Colv=TRUE, color=my_palette, scale="none", show_rownames=FALSE, show_colnames=FALSE, annotation_legend=TRUE, breaks=brk, annotation_colors = ann_colors, annotation_col = merged_types, annotation_names_col = FALSE, fontsize=12, treeheight_row = 20, treeheight_col = 20)
#dev.off()
```

```{r wgcna}
## TEIJ_20170125_RNA WGCNA

library(WGCNA)
library(tidyr)

#setwd("/Volumes/research/TEIJ_20170125_RNA/wgcna")
#load("/Volumes/research/TEIJ_20170125_RNA/wgcna/wgcna_data_and_tpm_figures.RData")

##Create CPM Matrix:

cpm <- cpm(y, log = TRUE, prior.count = 3)
cpm <- data.frame(cpm)
cpm_work=as.data.frame(t(cpm))

#checking to make sure samples are ok
goodgenescheck = goodSamplesGenes(cpm_work, verbose = 3)
goodgenescheck$allOK

### cluster samples to look at outliers 
hclust_tree = hclust(dist(cpm_work), method = "average")

## Plot the tree
par(cex = 0.6)
par(mar = c(0,4,2,0))
plot(hclust_tree, main = "Sample clustering to detect outliers", sub="", xlab="", cex.lab = 1.5,
     cex.axis = 1.5, cex.main = 2)

##Create a trait matrix:

traits <- as.data.frame(matrix(nrow = 16, ncol = 2))
colnames(traits) <- c("Saline", "MDMA")
rownames(traits) <- c("S1", "S2", "S3", "S4", "S5", "S6", "S7", "S8",
                      "MDMA1", "MDMA2", "MDMA3", "MDMA4", "MDMA5", "MDMA6", "MDMA7",
                      "MDMA9")
traits$Saline <- c(rep(1, 8), rep(0,8))
traits$MDMA <- c(rep(0, 8), rep(1,8))

# Re-cluster samples
hclust_tree2 = hclust(dist(cpm_work), method = "average")
# Convert traits to a color representation: blue means low, red means high, grey means missing entry
traitColors = numbers2colors(traits, signed = FALSE, colors=blueWhiteRed(100))
# Plot the sample dendrogram and the colors underneath.
plotDendroAndColors(hclust_tree2, traitColors,
                    groupLabels = names(traits),
                    main = "Sample dendrogram and trait heatmap")

#finding soft thresholding value
powers = c(c(1:10), seq(from = 12, to=50, by=2))
# Call the network topology analysis function
sft = pickSoftThreshold(cpm_work, powerVector = powers, verbose = 5)
# Plot the results:
par(mfrow = c(1,2))
cex1 = 0.9

# Scale-free topology fit index as a function of the soft-thresholding power
plot(sft$fitIndices[,1], -sign(sft$fitIndices[,3])*sft$fitIndices[,2],
     xlab="Soft Threshold (power)",ylab="Scale Free Topology Model Fit,signed R^2",type="n",
     main = paste("Scale independence"))
text(sft$fitIndices[,1], -sign(sft$fitIndices[,3])*sft$fitIndices[,2],
     labels=powers,cex=cex1,col="red");
# this line corresponds to using an R^2 cut-off of h
abline(h=c(0.90,0.80),col=c("red","blue"))
# Mean connectivity as a function of the soft-thresholding power
plot(sft$fitIndices[,1], sft$fitIndices[,5],
     xlab="Soft Threshold (power)",ylab="Mean Connectivity", type="n",
     main = paste("Mean connectivity"))
text(sft$fitIndices[,1], sft$fitIndices[,5], labels=powers, cex=cex1,col="red")

##Looks good! How about a treecut of .25 and a power of 7?

net = blockwiseModules(cpm_work, randomSeed=768, power = 6,
                       TOMType = "unsigned", minModuleSize = 30,
                       reassignThreshold = 0, mergeCutHeight = 0.3,
                       numericLabels = TRUE, pamRespectsDendro = FALSE,
                       saveTOMs = F,verbose = 3)


#checking number of modules
table(net$colors)

# Convert labels to colors for plotting
colors = labels2colors(net$colors)
# Plot the dendrogram and the module colors underneath
plotDendroAndColors(net$dendrograms[[1]], colors[net$blockGenes[[1]]],
                    "Module colors",
                    dendroLabels = FALSE, hang = 0.03,
                    addGuide = TRUE, guideHang = 0.05)
table(colors)

genes = ncol(cpm_work)
samples = nrow(cpm_work)
# Recalculate MEs with color labels
MEs0 = moduleEigengenes(cpm_work, colors)$eigengenes
MEs = orderMEs(MEs0)
moduleTraitCor = cor(MEs, traits, use = "p");
moduleTraitPvalue = corPvalueStudent(moduleTraitCor, samples)

textMatrix = paste(signif(moduleTraitCor, 2), "\n(",signif(moduleTraitPvalue, 1), ")", sep = "");
dim(textMatrix) = dim(moduleTraitCor)
par(mar = c(6, 8.5, 3, 3));
# Display the correlation values within a heatmap plot
labeledHeatmap(Matrix = moduleTraitCor,
               xLabels = names(traits),
               yLabels = substring(names(MEs),3),
               ySymbols = names(MEs),
               colorLabels = FALSE,
               colors = blueWhiteRed(50),
               textMatrix = textMatrix,
               setStdMargins = FALSE,
               cex.text = 0.5,
               zlim = c(-1,1),
               main = paste("Module-trait relationships"))


geneInfo=data.frame(Gene=colnames(cpm_work),Module=colors)

#write.table(geneInfo, "~/Documents/VANJ_20161220_RNA/WGCNA/VANJ_20161220_RNA_genes_in_modules.txt", sep="\t", quote = F, row.names = F)

##Find modules:
temp_geno = as.data.frame(traits)
modNames = substring(names(MEs), 3)
geneModuleMembership = as.data.frame(cor(cpm_work, MEs, use = "p"));
MMPvalue = as.data.frame(corPvalueStudent(as.matrix(geneModuleMembership), samples))
names(geneModuleMembership) = paste("MM", modNames, sep="");
names(MMPvalue) = paste("p.MM", modNames, sep="");
geneTraitSignificance = as.data.frame(cor(cpm_work, temp_geno, use = "p"));
GSPvalue = as.data.frame(corPvalueStudent(as.matrix(geneTraitSignificance), samples));
names(geneTraitSignificance) = paste("GS.", names(temp_geno), sep="");
names(GSPvalue) = paste("p.GS.", names(temp_geno), sep="")

module = "blue"
column=match(module, modNames)
moduleGenes=colors==module
par(mfrow=c(1,1))
verboseScatterplot(abs(geneModuleMembership[moduleGenes,column]),abs(geneTraitSignificance[moduleGenes,1]),
                   xlab=paste("Module Membership in",module,"module"),
                   ylab="Gene Significance for Trait",
                   main=paste("Module membership vs. gene significance\n"),cex.main = 1.2, cex.lab = 1.2, cex.axis = 1.2, col = module)


module = "darkolivegreen"
column=match(module, modNames)
moduleGenes=colors==module
par(mfrow=c(1,1))
verboseScatterplot(abs(geneModuleMembership[moduleGenes,column]),abs(geneTraitSignificance[moduleGenes,1]),
                   xlab=paste("Module Membership in",module,"module"),
                   ylab="Gene Significance for Trait",
                   main=paste("Module membership vs. gene significance\n"),cex.main = 1.2, cex.lab = 1.2, cex.axis = 1.2, col = module)


module = "turquoise"
column=match(module, modNames)
moduleGenes=colors==module
par(mfrow=c(1,1))
verboseScatterplot(abs(geneModuleMembership[moduleGenes,column]),abs(geneTraitSignificance[moduleGenes,1]),
                   xlab=paste("Module Membership in",module,"module"),
                   ylab="Gene Significance for Trait",
                   main=paste("Module membership vs. gene significance\n"),cex.main = 1.2, cex.lab = 1.2, cex.axis = 1.2, col = module)


module = "cyan"
column=match(module, modNames)
moduleGenes=colors==module
par(mfrow=c(1,1))
verboseScatterplot(abs(geneModuleMembership[moduleGenes,column]),abs(geneTraitSignificance[moduleGenes,1]),
                   xlab=paste("Module Membership in",module,"module"),
                   ylab="Gene Significance for Trait",
                   main=paste("Module membership vs. gene significance\n"),cex.main = 1.2, cex.lab = 1.2, cex.axis = 1.2, col = module)



module = "yellow"
column=match(module, modNames)
moduleGenes=colors==module
par(mfrow=c(1,1))
verboseScatterplot(abs(geneModuleMembership[moduleGenes,column]),abs(geneTraitSignificance[moduleGenes,1]),
                   xlab=paste("Module Membership in",module,"module"),
                   ylab="Gene Significance for Trait",
                   main=paste("Module membership vs. gene significance\n"),cex.main = 1.2, cex.lab = 1.2, cex.axis = 1.2, col = module)


module = "saddlebrown"
column=match(module, modNames)
moduleGenes=colors==module
par(mfrow=c(1,1))
verboseScatterplot(abs(geneModuleMembership[moduleGenes,column]),abs(geneTraitSignificance[moduleGenes,1]),
                   xlab=paste("Module Membership in",module,"module"),
                   ylab="Gene Significance for Trait",
                   main=paste("Module membership vs. gene significance\n"),cex.main = 1.2, cex.lab = 1.2, cex.axis = 1.2, col = module)

module = "brown"
column=match(module, modNames)
moduleGenes=colors==module
par(mfrow=c(1,1))
verboseScatterplot(abs(geneModuleMembership[moduleGenes,column]),abs(geneTraitSignificance[moduleGenes,1]),
                   xlab=paste("Module Membership in",module,"module"),
                   ylab="Gene Significance for Trait",
                   main=paste("Module membership vs. gene significance\n"),cex.main = 1.2, cex.lab = 1.2, cex.axis = 1.2, col = module)

####Finding hub genes

#blue
FilterGenes= abs(geneModuleMembership$MMblue)>.990
table(FilterGenes)
blue_hub=substring(dimnames(data.frame(cpm_work))[[2]][FilterGenes],1)
blue_hub <- as.data.frame(blue_hub)
colnames(blue_hub) <- "ens_gene"
blue_hub <- plyr::join(blue_hub, t2g, by = "ens_gene")
blue_hub <- subset(blue_hub, !duplicated(blue_hub$ens_gene))
blue_hub.sigdiff <- plyr::join(blue_hub, tt.annot, by = "ens_gene")
blue_hub.sigdiff <- blue_hub.sigdiff[,-c(11:14)]

write.table(blue_hub, "/Volumes/projects_secondary/bbc/research/BERA_20170817_RNA/analysis/edgeR/WGCNA/blue_hub_genes.txt", sep="\t", row.names=F, col.names=F, quote=F)

write.table(blue_hub.sigdiff, "/Volumes/projects_secondary/bbc/research/BERA_20170817_RNA/analysis/edgeR/WGCNA/blue_hub_genes_sigDGE.txt", sep="\t", row.names=F, col.names=F, quote=F)

#yellow
FilterGenes= abs(geneModuleMembership$MMyellow)>.930
table(FilterGenes)
yellow_hub=substring(dimnames(data.frame(cpm_work))[[2]][FilterGenes],1)
yellow_hub <- as.data.frame(yellow_hub)
colnames(yellow_hub) <- "ens_gene"
yellow_hub <- plyr::join(yellow_hub, t2g, by = "ens_gene")
yellow_hub <- subset(yellow_hub, !duplicated(yellow_hub$ens_gene))
yellow_hub.sigdiff <- plyr::join(yellow_hub, tt.annot, by = "ens_gene")
yellow_hub.sigdiff <- yellow_hub.sigdiff[,-c(11:14)] %>% subset(., complete.cases(.[,10]))

write.table(yellow_hub, "/Volumes/projects_secondary/bbc/research/BERA_20170817_RNA/analysis/edgeR/WGCNA/yellow_hub_genes.txt", sep="\t", row.names=F, col.names=F, quote=F)

write.table(yellow_hub.sigdiff, "/Volumes/projects_secondary/bbc/research/BERA_20170817_RNA/analysis/edgeR/WGCNA/yellow_hub_genes_sigDGE.txt", sep="\t", row.names=F, col.names=F, quote=F)

#darkolivegreen
FilterGenes= abs(geneModuleMembership$MMdarkolivegreen)>.820
table(FilterGenes)
green_hub=substring(dimnames(data.frame(cpm_work))[[2]][FilterGenes],1)
green_hub <- as.data.frame(green_hub)
colnames(green_hub) <- "ens_gene"
green_hub <- plyr::join(green_hub, t2g, by = "ens_gene")
green_hub <- subset(green_hub, !duplicated(green_hub$ens_gene))
green_hub.sigdiff <- plyr::join(green_hub, tt.annot, by = "ens_gene")
green_hub.sigdiff <- green_hub.sigdiff[,-c(11:14)] %>% subset(., complete.cases(.[,10]))

write.table(green_hub, "/Volumes/projects_secondary/bbc/research/BERA_20170817_RNA/analysis/edgeR/WGCNA/darkolivegreen_hub_genes.txt", sep="\t", row.names=F, col.names=F, quote=F)

write.table(green_hub.sigdiff, "/Volumes/projects_secondary/bbc/research/BERA_20170817_RNA/analysis/edgeR/WGCNA/darkolivegreen_hub_genes_sigDGE.txt", sep="\t", row.names=F, col.names=F, quote=F)

#turquoise
FilterGenes= abs(geneModuleMembership$MMturquoise)>.950
table(FilterGenes)
turquoise_hub=substring(dimnames(data.frame(cpm_work))[[2]][FilterGenes],1)
turquoise_hub <- as.data.frame(turquoise_hub)
colnames(turquoise_hub) <- "ens_gene"
turquoise_hub <- plyr::join(turquoise_hub, t2g, by = "ens_gene")
turquoise_hub <- subset(turquoise_hub, !duplicated(turquoise_hub$ens_gene))
turquoise_hub.sigdiff <- plyr::join(turquoise_hub, tt.annot, by = "ens_gene")
turquoise_hub.sigdiff <- turquoise_hub.sigdiff[,-c(11:14)] %>% subset(., complete.cases(.[,10]))

write.table(turquoise_hub, "/Volumes/projects_secondary/bbc/research/BERA_20170817_RNA/analysis/edgeR/WGCNA/turquoise_hub_genes.txt", sep="\t", row.names=F, col.names=F, quote=F)

write.table(turquoise_hub.sigdiff, "/Volumes/projects_secondary/bbc/research/BERA_20170817_RNA/analysis/edgeR/WGCNA/turquoise_hub_genes_sigDGE.txt", sep="\t", row.names=F, col.names=F, quote=F)

#brown
FilterGenes= abs(geneModuleMembership$MMbrown)>.90
table(FilterGenes)
brown_hub=substring(dimnames(data.frame(cpm_work))[[2]][FilterGenes],1)
brown_hub <- as.data.frame(brown_hub)
colnames(brown_hub) <- "ens_gene"
brown_hub <- plyr::join(brown_hub, t2g, by = "ens_gene")
brown_hub <- subset(brown_hub, !duplicated(brown_hub$ens_gene))
brown_hub.sigdiff <- plyr::join(brown_hub, tt.annot, by = "ens_gene")
brown_hub.sigdiff <- brown_hub.sigdiff[,-c(11:14)] %>% subset(., complete.cases(.[,10]))

write.table(brown_hub, "/Volumes/projects_secondary/bbc/research/BERA_20170817_RNA/analysis/edgeR/WGCNA/brown_hub_genes.txt", sep="\t", row.names=F, col.names=F, quote=F)

write.table(brown_hub.sigdiff, "/Volumes/projects_secondary/bbc/research/BERA_20170817_RNA/analysis/edgeR/WGCNA/brown_hub_genes_sigDGE.txt", sep="\t", row.names=F, col.names=F, quote=F)

#saddlebrown
FilterGenes= abs(geneModuleMembership$MMsaddlebrown)>.820
table(FilterGenes)
saddlebrown_hub=substring(dimnames(data.frame(cpm_work))[[2]][FilterGenes],1)
saddlebrown_hub <- as.data.frame(saddlebrown_hub)
colnames(saddlebrown_hub) <- "ens_gene"
saddlebrown_hub <- plyr::join(saddlebrown_hub, t2g, by = "ens_gene")
saddlebrown_hub <- subset(saddlebrown_hub, !duplicated(saddlebrown_hub$ens_gene))
saddlebrown_hub.sigdiff <- plyr::join(saddlebrown_hub, tt.annot, by = "ens_gene")
saddlebrown_hub.sigdiff <- saddlebrown_hub.sigdiff[,-c(11:14)] %>% subset(., complete.cases(.[,10]))

write.table(saddlebrown_hub, "/Volumes/projects_secondary/bbc/research/BERA_20170817_RNA/analysis/edgeR/WGCNA/saddlebrown_hub_genes.txt", sep="\t", row.names=F, col.names=F, quote=F)

write.table(saddlebrown_hub.sigdiff, "/Volumes/projects_secondary/bbc/research/BERA_20170817_RNA/analysis/edgeR/WGCNA/saddlebrown_hub_genes_sigDGE.txt", sep="\t", row.names=F, col.names=F, quote=F)

#cyan
FilterGenes= abs(geneModuleMembership$MMcyan)>.820
table(FilterGenes)
cyan_hub=substring(dimnames(data.frame(cpm_work))[[2]][FilterGenes],1)
cyan_hub <- as.data.frame(cyan_hub)
colnames(cyan_hub) <- "ens_gene"
cyan_hub <- plyr::join(cyan_hub, t2g, by = "ens_gene")
cyan_hub <- subset(cyan_hub, !duplicated(cyan_hub$ens_gene))
cyan_hub.sigdiff <- plyr::join(cyan_hub, tt.annot, by = "ens_gene")
cyan_hub.sigdiff <- cyan_hub.sigdiff[,-c(11:14)] %>% subset(., complete.cases(.[,10]))

write.table(cyan_hub, "/Volumes/projects_secondary/bbc/research/BERA_20170817_RNA/analysis/edgeR/WGCNA/cyan_hub_genes.txt", sep="\t", row.names=F, col.names=F, quote=F)

write.table(cyan_hub.sigdiff, "/Volumes/projects_secondary/bbc/research/BERA_20170817_RNA/analysis/edgeR/WGCNA/cyan_hub_genes_sigDGE.txt", sep="\t", row.names=F, col.names=F, quote=F)

####print out genes in each module

geneInfo=data.frame(Gene=colnames(cpm_work),Module=colors)

blue <- subset(geneInfo, Module =="blue")
bluegenes <- as.data.frame(blue$Gene)
colnames(bluegenes) <- c("Gene")
write.table(bluegenes, "/Volumes/projects_secondary/bbc/research/BERA_20170817_RNA/analysis/edgeR/WGCNA/blue_module_genes.txt", sep="\t", row.names=F, col.names=F, quote=F)

yellow <- subset(geneInfo, Module =="yellow")
yellowgenes <- as.data.frame(yellow$Gene)
colnames(yellowgenes) <- c("Gene")
write.table(yellowgenes, "/Volumes/projects_secondary/bbc/research/BERA_20170817_RNA/analysis/edgeR/WGCNA/yellow_module_genes.txt", sep="\t", row.names=F, col.names=F, quote=F)

cyan <- subset(geneInfo, Module =="cyan")
cyangenes <- as.data.frame(cyan$Gene)
colnames(cyangenes) <- c("Gene")
write.table(cyangenes, "/Volumes/projects_secondary/bbc/research/BERA_20170817_RNA/analysis/edgeR/WGCNA/cyan_module_genes.txt", sep="\t", row.names=F, col.names=F, quote=F)

brown <- subset(geneInfo, Module =="brown")
browngenes <- as.data.frame(brown$Gene)
colnames(browngenes) <- c("Gene")
write.table(browngenes, "/Volumes/projects_secondary/bbc/research/BERA_20170817_RNA/analysis/edgeR/WGCNA/brown_module_genes.txt", sep="\t", row.names=F, col.names=F, quote=F)

turquoise <- subset(geneInfo, Module =="turquoise")
turquoisegenes <- as.data.frame(turquoise$Gene)
colnames(turquoisegenes) <- c("Gene")
write.table(turquoisegenes, "/Volumes/projects_secondary/bbc/research/BERA_20170817_RNA/analysis/edgeR/WGCNA/turquoise_module_genes.txt", sep="\t", row.names=F, col.names=F, quote=F)

saddlebrown <- subset(geneInfo, Module =="saddlebrown")
saddlebrowngenes <- as.data.frame(saddlebrown$Gene)
colnames(saddlebrowngenes) <- c("Gene")
write.table(saddlebrowngenes, "/Volumes/projects_secondary/bbc/research/BERA_20170817_RNA/analysis/edgeR/WGCNA/saddlebrown_module_genes.txt", sep="\t", row.names=F, col.names=F, quote=F)

darkolivegreen <- subset(geneInfo, Module =="darkolivegreen")
darkolivegreengenes <- as.data.frame(darkolivegreen$Gene)
colnames(darkolivegreengenes) <- c("Gene")
write.table(darkolivegreengenes, "/Volumes/projects_secondary/bbc/research/BERA_20170817_RNA/analysis/edgeR/WGCNA/darkolivegreen_module_genes.txt", sep="\t", row.names=F, col.names=F, quote=F)


#Z-Score Plots
#I need to make expression matrices for each module
cpm$Gene <- rownames(cpm)
cpm <- cpm[,c(17, 1:16)]
rownames(cpm) <- NULL
colnames(cpm) <- c("Gene", paste0(rep("Saline_", 8), seq(1:8)),
                   paste0(rep("MDMA_", 8), seq(1:8)))

#make expression matrices 
se <- function(x) sqrt(rowVars(x)/length(x))

library(matrixStats)
blue_Salinese <- as.data.frame(se(as.matrix(cpm[,2:9])))
colnames(blue_Salinese) <- "se"
blue_Salinese$geno <- "Saline"

blue_MDMAse <- as.data.frame(se(as.matrix(cpm[,10:17])))
colnames(blue_MDMAse) <- "se"
blue_MDMAse$geno <- "MDMA"

blue_se <- rbind(blue_Salinese, blue_MDMAse)
blue_se <- as.data.frame(blue_se)

ggplot(blue_se, aes(se, fill = geno)) + geom_histogram(alpha = 0.2, bins = 100)
#se close to equal - pool

#blue
blue_matrix <- as.data.frame(plyr::join(cpm,bluegenes, by = "Gene", type = "inner"))
blue_Salinesd <- rowSds(as.matrix(blue_matrix[,2:9]))
blue_MDMAsd <- rowSds(as.matrix(blue_matrix[,10:17]))

blue_sd <- cbind(blue_Salinesd, blue_MDMAsd)
blue_sd <- as.data.frame(blue_sd)
blue_matrix$Salinem <- rowMeans(blue_matrix[,2:9])
blue_matrix$MDMAm <- rowMeans(blue_matrix[,10:17])

blue_zscores <- sapply(blue_matrix[,19], function(x) (x-blue_matrix$Salinem)/(sqrt((2 * (blue_sd$blue_WTsd^2 + blue_sd$blue_drp1sd^2 + blue_sd$blue_fzo1sd^2 + blue_sd$blue_eat3sd^2)) / 8)))
blue_zscores <- as.data.frame(blue_zscores)
blue_zscores$Gene <- blue_matrix$Gene
#blue_zscores$WTm <- scale(blue_matrix[, 2:4])
blue_zscores <- blue_zscores[, c(5, 1:4)]
write.table(blue_zscores, "~/Documents/VANJ_20161220_RNA/WGCNA/blue_module_zscore_expression_matrix_pooledvar.txt", row.names=F, col.names=F, quote=F,sep ="\t")

yellow_matrix <- as.data.frame(plyr::join(cpm, yellowgenes, by = "Gene", type = "inner"))
yellow_WTsd <- rowSds(as.matrix(yellow_matrix[,2:4]))
yellow_drp1sd <- rowSds(as.matrix(yellow_matrix[,5:7]))
yellow_fzo1sd <- rowSds(as.matrix(yellow_matrix[,8:10]))
yellow_eat3sd <- rowSds(as.matrix(yellow_matrix[,11:13]))
yellow_sd <- cbind(yellow_WTsd, yellow_drp1sd, yellow_fzo1sd, yellow_eat3sd)
yellow_sd <- as.data.frame(yellow_sd)
yellow_matrix$WTm <- rowMeans(yellow_matrix[,2:4])
yellow_matrix$drp1m <- rowMeans(yellow_matrix[,5:7])
yellow_matrix$fzo1m <- rowMeans(yellow_matrix[,8:10])
yellow_matrix$eat3m <- rowMeans(yellow_matrix[,11:13])

yellow_zscores <- sapply(yellow_matrix[,14:17], function(x) (x-yellow_matrix$WTm)/(sqrt((2 * (yellow_sd$yellow_WTsd^2 + yellow_sd$yellow_drp1sd^2 + yellow_sd$yellow_fzo1sd^2 + yellow_sd$yellow_eat3sd^2)) / 8)))
yellow_zscores <- as.data.frame(yellow_zscores)
yellow_zscores$Gene <- yellow_matrix$Gene
#blue_zscores$WTm <- scale(blue_matrix[, 2:4])
yellow_zscores <- yellow_zscores[, c(5, 1:4)]
write.table(yellow_zscores, "~/Documents/VANJ_20161220_RNA/WGCNA/yellow_module_zscore_expression_matrix_pooledvar.txt", row.names=F, col.names=F, quote=F,sep ="\t")

green_matrix <- as.data.frame(plyr::join(cpm, greengenes, by = "Gene", type = "inner"))
green_WTsd <- rowSds(as.matrix(green_matrix[,2:4]))
green_drp1sd <- rowSds(as.matrix(green_matrix[,5:7]))
green_fzo1sd <- rowSds(as.matrix(green_matrix[,8:10]))
green_eat3sd <- rowSds(as.matrix(green_matrix[,11:13]))
green_sd <- cbind(green_WTsd, green_drp1sd, green_fzo1sd, green_eat3sd)
green_sd <- as.data.frame(green_sd)
green_matrix$WTm <- rowMeans(green_matrix[,2:4])
green_matrix$drp1m <- rowMeans(green_matrix[,5:7])
green_matrix$fzo1m <- rowMeans(green_matrix[,8:10])
green_matrix$eat3m <- rowMeans(green_matrix[,11:13])

green_zscores <- sapply(green_matrix[,14:17], function(x) (x-green_matrix$WTm)/(sqrt((2 * (green_sd$green_WTsd^2 + green_sd$green_drp1sd^2 + green_sd$green_fzo1sd^2 + green_sd$green_eat3sd^2)) / 8)))
green_zscores <- as.data.frame(green_zscores)
green_zscores$Gene <- green_matrix$Gene
#blue_zscores$WTm <- scale(blue_matrix[, 2:4])
green_zscores <- green_zscores[, c(5, 1:4)]
write.table(green_zscores, "~/Documents/VANJ_20161220_RNA/WGCNA/green_module_zscore_expression_matrix_pooledvar.txt", row.names=F, col.names=F, quote=F,sep ="\t")

brown_matrix <- as.data.frame(plyr::join(cpm, browngenes, by = "Gene", type = "inner"))
brown_WTsd <- rowSds(as.matrix(brown_matrix[,2:4]))
brown_drp1sd <- rowSds(as.matrix(brown_matrix[,5:7]))
brown_fzo1sd <- rowSds(as.matrix(brown_matrix[,8:10]))
brown_eat3sd <- rowSds(as.matrix(brown_matrix[,11:13]))
brown_sd <- cbind(brown_WTsd, brown_drp1sd, brown_fzo1sd, brown_eat3sd)
brown_sd <- as.data.frame(brown_sd)
brown_matrix$WTm <- rowMeans(brown_matrix[,2:4])
brown_matrix$drp1m <- rowMeans(brown_matrix[,5:7])
brown_matrix$fzo1m <- rowMeans(brown_matrix[,8:10])
brown_matrix$eat3m <- rowMeans(brown_matrix[,11:13])

brown_zscores <- sapply(brown_matrix[,14:17], function(x) (x-brown_matrix$WTm)/(sqrt((2 * (brown_sd$brown_WTsd^2 + brown_sd$brown_drp1sd^2 + brown_sd$brown_fzo1sd^2 + brown_sd$brown_eat3sd^2)) / 8)))
brown_zscores <- as.data.frame(brown_zscores)
brown_zscores$Gene <- brown_matrix$Gene
#blue_zscores$WTm <- scale(blue_matrix[, 2:4])
brown_zscores <- brown_zscores[, c(5, 1:4)]
write.table(brown_zscores, "~/Documents/VANJ_20161220_RNA/WGCNA/brown_module_zscore_expression_matrix_pooledvar.txt", row.names=F, col.names=F, quote=F,sep ="\t")

turquoise_matrix <- as.data.frame(plyr::join(cpm, turquoisegenes, by = "Gene", type = "inner"))
turquoise_WTsd <- rowSds(as.matrix(turquoise_matrix[,2:4]))
turquoise_drp1sd <- rowSds(as.matrix(turquoise_matrix[,5:7]))
turquoise_fzo1sd <- rowSds(as.matrix(turquoise_matrix[,8:10]))
turquoise_eat3sd <- rowSds(as.matrix(turquoise_matrix[,11:13]))
turquoise_sd <- cbind(turquoise_WTsd, turquoise_drp1sd, turquoise_fzo1sd, turquoise_eat3sd)
turquoise_sd <- as.data.frame(turquoise_sd)
turquoise_matrix$WTm <- rowMeans(turquoise_matrix[,2:4])
turquoise_matrix$drp1m <- rowMeans(turquoise_matrix[,5:7])
turquoise_matrix$fzo1m <- rowMeans(turquoise_matrix[,8:10])
turquoise_matrix$eat3m <- rowMeans(turquoise_matrix[,11:13])

turquoise_zscores <- sapply(turquoise_matrix[,14:17], function(x) (x-turquoise_matrix$WTm)/(sqrt((2 * (turquoise_sd$turquoise_WTsd^2 + turquoise_sd$turquoise_drp1sd^2 + turquoise_sd$turquoise_fzo1sd^2 + turquoise_sd$turquoise_eat3sd^2)) / 8)))
turquoise_zscores <- as.data.frame(turquoise_zscores)
turquoise_zscores$Gene <- turquoise_matrix$Gene
#blue_zscores$WTm <- scale(blue_matrix[, 2:4])
turquoise_zscores <- turquoise_zscores[, c(5, 1:4)]
write.table(turquoise_zscores, "~/Documents/VANJ_20161220_RNA/WGCNA/turquoise_module_zscore_expression_matrix_pooledvar.txt", row.names=F, col.names=F, quote=F,sep ="\t")

red_matrix <- as.data.frame(plyr::join(cpm, redgenes, by = "Gene", type = "inner"))
red_WTsd <- rowSds(as.matrix(red_matrix[,2:4]))
red_drp1sd <- rowSds(as.matrix(red_matrix[,5:7]))
red_fzo1sd <- rowSds(as.matrix(red_matrix[,8:10]))
red_eat3sd <- rowSds(as.matrix(red_matrix[,11:13]))
red_sd <- cbind(red_WTsd, red_drp1sd, red_fzo1sd, red_eat3sd)
red_sd <- as.data.frame(red_sd)
red_matrix$WTm <- rowMeans(red_matrix[,2:4])
red_matrix$drp1m <- rowMeans(red_matrix[,5:7])
red_matrix$fzo1m <- rowMeans(red_matrix[,8:10])
red_matrix$eat3m <- rowMeans(red_matrix[,11:13])

red_zscores <- sapply(red_matrix[,14:17], function(x) (x-red_matrix$WTm)/(sqrt((2 * (red_sd$red_WTsd^2 + red_sd$red_drp1sd^2 + red_sd$red_fzo1sd^2 + red_sd$red_eat3sd^2)) / 8)))
red_zscores <- as.data.frame(red_zscores)
red_zscores$Gene <- red_matrix$Gene
#blue_zscores$WTm <- scale(blue_matrix[, 2:4])
red_zscores <- red_zscores[, c(5, 1:4)]
write.table(red_zscores, "~/Documents/VANJ_20161220_RNA/WGCNA/red_module_zscore_expression_matrix_pooledvar.txt", row.names=F, col.names=F, quote=F,sep ="\t")

#Z Score Figures
blue_zscores.hubs <- subset(blue_zscores, blue_zscores$Gene %in% blue_hub)
colnames(blue_zscores.hubs) <- c("Gene", "WT", "drp-1", "fzo-1", "eat-3")
file.m.blue <- melt(blue_zscores.hubs[,c(1,3:5)], id.vars = c("Gene"))

turquoise_zscores.hubs <- subset(turquoise_zscores, turquoise_zscores$Gene %in% turquoise_hub)
colnames(turquoise_zscores.hubs) <- c("Gene", "WT", "drp-1", "fzo-1", "eat-3")
file.m.turquoise <- melt(turquoise_zscores.hubs[,c(1,3:5)], id.vars = c("Gene"))

brown_zscores.hubs <- subset(brown_zscores, brown_zscores$Gene %in% brown_hub)
colnames(brown_zscores.hubs) <- c("Gene", "WT", "drp-1", "fzo-1", "eat-3")
file.m.brown <- melt(brown_zscores.hubs[,c(1,3:5)], id.vars = c("Gene"))

green_zscores.hubs <- subset(green_zscores, green_zscores$Gene %in% green_hub)
colnames(green_zscores.hubs) <- c("Gene", "WT", "drp-1", "fzo-1", "eat-3")
file.m.green <- melt(green_zscores.hubs[,c(1,3:5)], id.vars = c("Gene"))

yellow_zscores.hubs <- subset(yellow_zscores, yellow_zscores$Gene %in% yellow_hub)
colnames(yellow_zscores.hubs) <- c("Gene", "WT", "drp-1", "fzo-1", "eat-3")
file.m.yellow <- melt(yellow_zscores.hubs[,c(1,3:5)], id.vars = c("Gene"))

red_zscores.hubs <- subset(red_zscores, red_zscores$Gene %in% red_hub)
colnames(red_zscores.hubs) <- c("Gene", "WT", "drp-1", "fzo-1", "eat-3")
file.m.red <- melt(red_zscores.hubs[,c(1,3:5)], id.vars = c("Gene"))

p1 <- ggplot(file.m.brown, aes(variable, value, group=factor(Gene))) + 
  geom_line(aes(color=factor(Gene))) +
  ggtitle("Brown") +
  theme_classic(12) +
  theme(
    legend.position="none",
    axis.line.x=element_line(size=.7528125,color="black"),
    axis.line.y=element_line(size=.7528125,color="black"),
    axis.ticks.length = unit(.2,"cm"),
    axis.title.x = element_blank(),
    axis.title.y = element_text(size=14,face="bold",family="Arial",margin=margin(0,.2,0,0,unit="in"),vjust=.5,hjust=.5),
    axis.text.y =  element_text(size=12,face="bold",family="Arial",margin=margin(0,0,0,0,unit="in"),vjust=.5,hjust=1),
    axis.text.x = element_text(size=12,face="bold",family="Arial", vjust=.6,angle=45),
    axis.ticks=element_line(size=.75),
    legend.key.width = unit(1, "lines"),
    title = element_text(size=12,vjust=2,face="bold",family="Arial"),
    plot.title = element_text(hjust = 0.5)) +
  ylab("zscore") +
  ylim(-3.0,8.0)

p2 <- ggplot(file.m.turquoise, aes(variable, value, group=factor(Gene))) + 
  geom_line(aes(color=factor(Gene))) +
  ggtitle("Turquoise") +
  theme_classic(12) +
  theme(
    legend.position="none",
    axis.line.x=element_line(size=.7528125,color="black"),
    axis.line.y=element_line(size=.7528125,color="black"),
    axis.ticks.length = unit(.2,"cm"),
    axis.title.x = element_blank(),
    axis.title.y = element_text(size=14,face="bold",family="Arial",margin=margin(0,.2,0,0,unit="in"),vjust=.5,hjust=.5),
    axis.text.y =  element_text(size=12,face="bold",family="Arial",margin=margin(0,0,0,0,unit="in"),vjust=.5,hjust=1),
    axis.text.x = element_text(size=12,face="bold",family="Arial", vjust=.6,angle=45),
    axis.ticks=element_line(size=.75),
    legend.key.width = unit(1, "lines"),
    title = element_text(size=12,vjust=2,face="bold",family="Arial"),
    plot.title = element_text(hjust = 0.5)) +
  ylab("zscore") +
  ylim(-14.0,12.0)

p3 <- ggplot(file.m.green, aes(variable, value, group=factor(Gene))) + 
  geom_line(aes(color=factor(Gene))) +
  ggtitle("Green") +
  theme_classic(12) +
  theme(
    legend.position="none",
    axis.line.x=element_line(size=.7528125,color="black"),
    axis.line.y=element_line(size=.7528125,color="black"),
    axis.ticks.length = unit(.2,"cm"),
    axis.title.x = element_blank(),
    axis.title.y = element_text(size=14,face="bold",family="Arial",margin=margin(0,.2,0,0,unit="in"),vjust=.5,hjust=.5),
    axis.text.y =  element_text(size=12,face="bold",family="Arial",margin=margin(0,0,0,0,unit="in"),vjust=.5,hjust=1),
    axis.text.x = element_text(size=12,face="bold",family="Arial", vjust=.6,angle=45),
    axis.ticks=element_line(size=.75),
    legend.key.width = unit(1, "lines"),
    title = element_text(size=12,vjust=2,face="bold",family="Arial"),
    plot.title = element_text(hjust = 0.5)) +
  ylab("zscore") +
  ylim(-7.0,10.0)

p4 <- ggplot(file.m.yellow, aes(variable, value, group=factor(Gene))) + 
  geom_line(aes(color=factor(Gene))) +
  ggtitle("Yellow") +
  theme_classic(12) +
  theme(
    legend.position="none",
    axis.line.x=element_line(size=.7528125,color="black"),
    axis.line.y=element_line(size=.7528125,color="black"),
    axis.ticks.length = unit(.2,"cm"),
    axis.title.x = element_blank(),
    axis.title.y = element_text(size=14,face="bold",family="Arial",margin=margin(0,.2,0,0,unit="in"),vjust=.5,hjust=.5),
    axis.text.y =  element_text(size=12,face="bold",family="Arial",margin=margin(0,0,0,0,unit="in"),vjust=.5,hjust=1),
    axis.text.x = element_text(size=12,face="bold",family="Arial", vjust=.6,angle=45),
    axis.ticks=element_line(size=.75),
    legend.key.width = unit(1, "lines"),
    title = element_text(size=12,vjust=2,face="bold",family="Arial"),
    plot.title = element_text(hjust = 0.5)) +
  ylab("zscore") +
  ylim(-7.0,3.0)

p5 <- ggplot(file.m.blue, aes(variable, value, group=factor(Gene))) + 
  geom_line(aes(color=factor(Gene))) +
  ggtitle("Blue") +
  theme_classic(12) +
  theme(
    legend.position="none",
    axis.line.x=element_line(size=.7528125,color="black"),
    axis.line.y=element_line(size=.7528125,color="black"),
    axis.ticks.length = unit(.2,"cm"),
    axis.title.x = element_blank(),
    axis.title.y = element_text(size=14,face="bold",family="Arial",margin=margin(0,.2,0,0,unit="in"),vjust=.5,hjust=.5),
    axis.text.y =  element_text(size=12,face="bold",family="Arial",margin=margin(0,0,0,0,unit="in"),vjust=.5,hjust=1),
    axis.text.x = element_text(size=12,face="bold",family="Arial", vjust=.6,angle=45),
    axis.ticks=element_line(size=.75),
    legend.key.width = unit(1, "lines"),
    title = element_text(size=12,vjust=2,face="bold",family="Arial"),
    plot.title = element_text(hjust = 0.5)) +
  ylab("zscore") +
  ylim(-3.0,13.0)

p6 <- ggplot(file.m.red, aes(variable, value, group=factor(Gene))) + 
  geom_line(aes(color=factor(Gene))) +
  ggtitle("Red") +
  theme_classic(12) +
  theme(
    legend.position="none",
    axis.line.x=element_line(size=.7528125,color="black"),
    axis.line.y=element_line(size=.7528125,color="black"),
    axis.ticks.length = unit(.2,"cm"),
    axis.title.x = element_blank(),
    axis.title.y = element_text(size=14,face="bold",family="Arial",margin=margin(0,.2,0,0,unit="in"),vjust=.5,hjust=.5),
    axis.text.y =  element_text(size=12,face="bold",family="Arial",margin=margin(0,0,0,0,unit="in"),vjust=.5,hjust=1),
    axis.text.x = element_text(size=12,face="bold",family="Arial", vjust=.6,angle=45),
    axis.ticks=element_line(size=.75),
    legend.key.width = unit(1, "lines"),
    title = element_text(size=12,vjust=2,face="bold",family="Arial"),
    plot.title = element_text(hjust = 0.5)) +
  ylab("zscore") +
  ylim(-2.0,7.0)

pg <- plot_grid(p1, p2, p3, p4, p5, p6, align = "hv", nrow = 2, ncol = 3, labels = c("A", "B", "C", "D", "E", "F"))

###Module annotation:

library("biomaRt")
library("GO.db")

setwd("~/Documents/VANJ_20161220_RNA/WGCNA")

turquoise <- read.table("turquoise_module_genes.txt")
yellow <- read.table("yellow_module_genes.txt")
green <- read.table("green_module_genes.txt")
red <- read.table("red_module_genes.txt")
brown <- read.table("brown_module_genes.txt")
blue <- read.table("blue_module_genes.txt")

ensembl = useMart("ENSEMBL_MART_ENSEMBL", dataset="celegans_gene_ensembl",
                  host = "uswest.ensembl.org")

goAnnot_turquoise=getBM(c("ensembl_gene_id","external_gene_name", "description", "gene_biotype"),filters="ensembl_gene_id",values=turquoise,mart=ensembl)
write.table(goAnnot_turquoise, "goAnnot_turquoise_genes.txt", row.names=F, col.names=F, quote=F, sep ="\t")

goAnnot_yellow=getBM(c("ensembl_gene_id","external_gene_name","description", "gene_biotype"),filters="ensembl_gene_id",values=yellow,mart=ensembl)
write.table(goAnnot_yellow, "goAnnot_yellow_genes.txt", row.names=F, col.names=F, quote=F, sep ="\t")

goAnnot_brown=getBM(c("ensembl_gene_id","external_gene_name","description", "gene_biotype"),filters="ensembl_gene_id",values=brown,mart=ensembl)
write.table(goAnnot_brown, "goAnnot_brown_genes.txt", row.names=F, col.names=F, quote=F, sep ="\t")

goAnnot_blue=getBM(c("ensembl_gene_id","external_gene_name","description", "gene_biotype"),filters="ensembl_gene_id",values=blue,mart=ensembl)
write.table(goAnnot_blue, "goAnnot_blue_genes.txt", row.names=F, col.names=F, quote=F, sep ="\t")

goAnnot_green=getBM(c("ensembl_gene_id","external_gene_name","description", "gene_biotype"),filters="ensembl_gene_id",values=green,mart=ensembl)
write.table(goAnnot_green, "goAnnot_green_genes.txt", row.names=F, col.names=F, quote=F, sep ="\t")

goAnnot_red=getBM(c("ensembl_gene_id","external_gene_name","description", "gene_biotype"),filters="ensembl_gene_id",values=red,mart=ensembl)
write.table(goAnnot_red, "goAnnot_red_genes.txt", row.names=F, col.names=F, quote=F, sep ="\t")

#####Differential Expression

tt <- topTable(fit2, coef = 1, n=Inf, sort.by = "p")
tt.subset1 <- subset(tt, adj.P.Val < 0.05) # drp1 ~ WT
tt.subset1$ensembl_gene_id <- rownames(tt.subset1)
annot_tt.subset1=getBM(c("ensembl_gene_id","external_gene_name", "description", "gene_biotype"),filters="ensembl_gene_id",values=tt.subset1$ensembl_gene_id,mart=ensembl)
merge1 <- plyr::join(annot_tt.subset1,tt.subset1, by ="ensembl_gene_id")

write.table(merge1, "drp1_DE_genes_vs_WT_wgcna_limmavoom.txt", row.names=F, col.names=T, quote=F, sep ="\t")

tt1 <- topTable(fit2, coef=2, n=nrow(y$E), sort.by = "p")
tt.subset2 <- subset(tt1, adj.P.Val < 0.05) # fzo1 ~ WT
tt.subset2$ensembl_gene_id <- rownames(tt.subset2)
annot_tt.subset2=getBM(c("ensembl_gene_id","external_gene_name", "description", "gene_biotype"),filters="ensembl_gene_id",values=tt.subset2$ensembl_gene_id,mart=ensembl)
merge2 <- plyr::join(annot_tt.subset2,tt.subset2, by ="ensembl_gene_id")

write.table(merge2, "fzo1_DE_genes_vs_WT_wgcna_limmavoom.txt", row.names=F, col.names=T, quote=F, sep ="\t")

tt2 <- topTable(fit2, coef=3, n=nrow(y$E), sort.by = "p")
tt.subset3 <- subset(tt2, adj.P.Val < 0.05) # FusionMut ~ WT
tt.subset3$ensembl_gene_id <- rownames(tt.subset3)
annot_tt.subset3=getBM(c("ensembl_gene_id","external_gene_name", "description", "gene_biotype"),filters="ensembl_gene_id",values=tt.subset3$ensembl_gene_id,mart=ensembl)
merge3 <- plyr::join(annot_tt.subset3,tt.subset3, by ="ensembl_gene_id")

write.table(merge3, "eat3_DE_genes_vs_WT_wgcna_limmavoom.txt", row.names=F, col.names=T, quote=F, sep ="\t")


#Find genes that are only different FusionMut samples: 

join1 <- semi_join(tt.subset2, tt.subset3, by = "ensembl_gene_id")
#20 genes are sig diff in FusionMut relative to WT AND FissionMut

#Find genes that are only different FissionMut samples: 

fullgenes <- rbind(tt.subset1, tt.subset2, tt.subset3)
join2 <- fullgenes[!duplicated(fullgenes$ensembl_gene_id, fromLast=TRUE), ]
#218 genes are sig diff in FissionMut relative to WT AND FusionMut

#Annotate FusionMut specific genes:

join1_gene <- join1$ensembl_gene_id

annot_join1=getBM(c("ensembl_gene_id","external_gene_name", "description", "gene_biotype"),filters="ensembl_gene_id",values=join1_gene,mart=ensembl)

join2_gene <- join2$ensembl_gene_id

annot_join2=getBM(c("ensembl_gene_id","external_gene_name", "description", "gene_biotype"),filters="ensembl_gene_id",values=join2_gene,mart=ensembl)

#What modules are these genes in? 

modules <- read.table("VANJ_20161220_RNA_genes_in_modules.txt", sep="\t", header= T)
colnames(modules) <- c("ensembl_gene_id", "module")

mod_join1 <- inner_join(annot_join1, modules, by = "ensembl_gene_id")

mod_join2 <- inner_join(annot_join2, modules, by = "ensembl_gene_id")

mod_join3 <- inner_join(mod_join1,tt.subset1, by = "ensembl_gene_id")
write.table(mod_join3, "FusionMut_speciifc_DE_Genes.txt", row.names=F, col.names=T, quote=F, sep ="\t")

mod_join4 <- inner_join(mod_join2,tt.subset4, by = "ensembl_gene_id")
write.table(mod_join4, "FissionMut_speciifc_DE_Genes.txt", row.names=F, col.names=T, quote=F, sep ="\t")



#####
```
