---
title: "RNA-seq Analysis Pipeline"
author: "Joe Kochmanski"
date: 4/28/2022
output: html_document
---


## Differential transcript expression using sleuth pipeline (pseudoalignment to index transcriptome)

```{r Load required packages}
#### Pipeline for analysis of RNA-seq data ###

#Step 1
#Load required libraries:
source("https://bioconductor.org/biocLite.R")
BiocManager::install(c("tximport"))
BiocManager::install(c("devtools"))
BiocManager::install(c("COMBINE-lab/wasabi"))
BiocManager::install(c("remotes"))
BiocManager::install(c("pachterlab/sleuth"))
BiocManager::install(c("goseq"))
BiocManager::install(c("biomaRt"))
install.packages("cowplot")

library(tximport)
library(readr)
library(tidyr)
library(plyr)
library(devtools)
library(wasabi)
library(sleuth)
library(cowplot)
library(goseq)
library(biomaRt)
```

```{r Read in quantified RNA-seq data}
#Get the sample ids
sample_id <- dir(file.path("/Users/JoeKochmanski/Desktop/quants2/", "data_sleuth"))
sample_id #lists sample ids.

#Read in table of meta data that includes /path for each sample's wasabi-processed data. Only males this time:
meta.path.M <- read.table("/Users/JoeKochmanski/Desktop/quants2/RNAseq_AnnaStoll_OnlyMale_Meta_Wasabi_Path_060921.txt", header = TRUE)
print(meta.path.M)

#Side note: path must be character for running in sleuth:
meta.path.M$path <- as.character(meta.path.M$path)
meta.path.M$path
```

```{r Build databases for organism of interest (i.e. rat)}
#Build databases

# Set the host
# mart <- biomaRt::useMart(biomart = 'ENSEMBL_MART_ENSEMBL')
# mart <- biomaRt::useMart(biomart = 'ENSEMBL_MART_ENSEMBL'); View(listDatasets(mart))
mart <- biomaRt::useMart(biomart = 'ENSEMBL_MART_ENSEMBL', host="useast.ensembl.org",dataset = 'rnorvegicus_gene_ensembl') # this is the dataset (Mouse genes (GRCm38.p5) used by UCSC to create their mm10 database

t2g <- biomaRt::getBM(attributes = c("ensembl_transcript_id", "ensembl_gene_id", "external_gene_name", "description"), mart = mart, useCache = FALSE)

t2g <- dplyr::rename(t2g, ens_transcr = ensembl_transcript_id, ens_gene = ensembl_gene_id, ext_gene = external_gene_name, predicted_function = description)
head(t2g)

# Rename again for Sleuth
t2gSleuth <- dplyr::rename(t2g, target_id = ens_transcr) # target is the transcript
```

```{r Differential transcript expression (DTE) analysis}
############# DTE with Sleuth ##########################################

#Sleuth prep
full_design.M <- model.matrix(~meta.path.M$treatment) # NO COVARIATES; ONLY TREATMENT.

colnames(full_design.M)[1] <- c("Intercept") #Create variable for intercept.

#A basic filter to apply to sleuth prep -- 50% of samples must have at least 10 est counts
my_basic_filter <- function(row, min_reads = 10, min_prop = 0.5) {
  mean(row >= min_reads) >= min_prop
}

#The 'sleuth' object stores a group of wasabi results, and can then operate on them while accounting for covariates, sequencing depth, technical and biological variance.
so.M <- sleuth_prep(meta.path.M, full_model = full_design.M, target_mapping = t2gSleuth, filter_fun = my_basic_filter, extra_bootstrap_summary = TRUE)

#Fit the models
so.M <- sleuth_fit(so.M, fit_name="full") # full model, uses full_design to design model based on treatment.

#What this has accomplished is to “smooth” the raw kallisto abundance estimates for each sample using a linear model with a parameter that represents the experimental condition. 
#To test for transcripts that are differential expressed between the conditions, sleuth performs a second fit to a “reduced” model that presumes abundances are equal in the two conditions. 
#To identify differential expressed transcripts sleuth will then identify transcripts with a significantly better fit in the “full” model.
#Fit the reduced model(s). In this case the 'reduced' model is one with ONLY the intercept (null model):
so.M <- sleuth_fit(so.M, formula = ~1, fit_name = "reduced")

#View the models
models(so.M)
tests(so.M)

#Likelihood ratio test (DIFFERENTIAL TRANSCRIPT EXPRESSION):
so.M <- sleuth_lrt(so.M, "reduced", "full")
```

```{r Pull out differential testing results as data frame and filter by q-value cutoff}
#Let's make the results a data frame.
sleuth_table_gene.M <- sleuth_results(so.M, 'reduced:full', 'lrt', show_all = FALSE) 

#Now let's filter by q-value < 0.05
sleuth_table_gene.M.sig <- dplyr::filter(sleuth_table_gene.M, qval <= 0.05) # 495 observations 
head(sleuth_table_gene.M.sig)
```

```{r Run separate Wald Test to generate fold change estimate}
## Run the Wald test (WT), a statistical tests which:
# - is somewhat related to the LRT and is also used to test for differential expression
# - LRT is considered a better test than the WT but
# - WT is used because it generates the beta statistic, which approximates to the fold change in expression between
# the 2 condition tested, which is typically reported in differential expression analysis
so.M.wt <- sleuth_wt(so.M, "meta.path.M$treatment")

models(so.M.wt) #View available models
tests(so.M.wt) #View available tests

sleuth_table_gene.M.wt <- sleuth_results(so.M.wt, 'meta.path.M$treatment', test_type='wt', show_all = FALSE) 
head(sleuth_table_gene.M.wt)

#Now crunch down the wald test results to only the target ID, b, and se_b variables
myvars <- c("target_id", "b", "se_b")
sleuth_table_gene.M.wt.1 <- sleuth_table_gene.M.wt[myvars]

#Now merge together the LRT results with the Wald test variables of interest - beta and standard error
sleuth_table_gene.M.full <- merge(sleuth_table_gene.M, sleuth_table_gene.M.wt.1[, c('target_id', 'b', 'se_b')], on = 'target_id', sort = FALSE)

#Notes about beta estimates from Wald Test:
#Internally, with sleuth_prep Sleuth will transform elements in the condition field to 0s and 1s in alphabetical order 
#and then WT's beta values will be relative to the 0 condition; 
#that is, positive beta values showing transcripts in which expression is greater in condition 1 than in condition 0. 
#Based on the coding of the 'treatment' variable, beta values would be relative to the 'Control' samples, which are coded as 0. 'PFF' is coded as 1.
#If using a categorical treatment, make sure that the desired reference condition ranks first alphabetically speaking.
```

```{r View example of significant hits}
#The most significant genes are pulled out using the following command:
head(sleuth_table_gene.M.full, 20)

#DTE lists - Note: This is for Differential transcript testing, NOT gene-level testing.
#Because gene results are built on transcript results, the gene and transcript results are entirely consistent and compatible with each other. To visualize the gene-level results, one merely runs sleuth_results again but this time setting the flag ‘pval_aggregate’ to TRUE.

#sleuth_table_gene <- sleuth_results(so, 'reduced:full', 'lrt', show_all = FALSE, pval_aggregate = TRUE)
#sleuth_table_gene <- dplyr::filter(sleuth_table_tx, qval <= 0.05)
#head(sleuth_table_gene, 20)
#Female only:
#sleuth_table.F <- sleuth_results(so.F, test='reduced:full', test_type='lrt', show_all = FALSE, pval_aggregate = TRUE)
#sleuth_significant.F <- dplyr::filter(sleuth_table.F, qval <= 0.05) #0 observations w/ q-value < 0.05
#Male only:
#sleuth_table.M <- sleuth_results(so.M, test='reduced:full', test_type='lrt', show_all = FALSE, pval_aggregate = TRUE)
#sleuth_significant.M <- dplyr::filter(sleuth_table.M, qval <= 0.05) #0 observations w/ q-value < 0.05 
```

```{r Write results as separate .csv file and further examining results}

#Write sleuth test tables for stratified analysis (including non-significant results) to .csv files:
write.csv(sleuth_table_gene.M.full, "/Users/JoeKochmanski/Dropbox/Bernstein_Lab/Random TSMM Stuff/AnnaStoll_RNAseq/RNAseq_Sleuth_DGE_Male_TestTable_v3_060921.csv")

#Can also cut down to only those with qvalue < 0.2
sleuth_table_gene.M.full.q <- sleuth_table_gene.M.full[which(sleuth_table_gene.M.full$qval < 0.2),] #141 observations
#2534 significant transcripts

#go to sleuth live to view results in tabular form; useful R shiny interface for examining specific genes/transcripts.
sleuth_live(so.M)

#Turn off sleuth_live interface.
dev.off()

# Saving on object in RData format
save(so.M, file = "/Users/JoeKochmanski/Dropbox/Bernstein_Lab/Random TSMM Stuff/AnnaStoll_RNAseq/AnnaStoll_DGE_sleuth_object_male_LRT_060921.RData")
save(so.M.wt, file = "/Users/JoeKochmanski/Dropbox/Bernstein_Lab/Random TSMM Stuff/AnnaStoll_RNAseq/AnnaStoll_DGE_sleuth_object_male_WT_060921.RData")

###### End of DTE analysis ##########
```

```{r OPTIONAL: If multiple analyses performed, try comparing lists!}

##OPTIONAL:Try comparing the lists of differentially expressed genes for males and females:

#Read in list of male DGE results
male.dge <- read.csv("/Users/jjkoch/Dropbox/Bernstein_Lab/Random TSMM Stuff/AnnaStoll_RNAseq/RNAseq_Sleuth_DGE_Male_TestTable_v3_060921.csv")
head(male.dge)
#Now cut down to only those with qvalue < 0.2
male.dge.q <- male.dge[which(male.dge$qval < 0.2),] #2534 observations
male.dge.q.up <- male.dge.q[which(male.dge.q$b > 0),] #990 observations
male.dge.q.down <- male.dge.q[which(male.dge.q$b < 0),] #1544 observations

#Read in list of female DGE results
female.dge <- read.csv("/Users/jjkoch/Dropbox/Bernstein_Lab/Random TSMM Stuff/AnnaStoll_RNAseq/RNAseq_Sleuth_DGE_Female_TestTable_v2_040921.csv")
head(female.dge)
#Now cut down to only those with qvalue < 0.2
female.dge.q <- female.dge[which(female.dge$qval < 0.2),] #1419 observations
female.dge.q.up <- female.dge.q[which(female.dge.q$b > 0),] #1019 observations
female.dge.q.down <- female.dge.q[which(female.dge.q$b < 0),] #400 observations

#Test for overlap by 'ext_gene'
male.female.merge <- inner_join(male.dge.q, female.dge.q, by = "target_id") #326 transcripts.

#Which of these overlapping transcripts show the same direction:

male.female.merge.up <- male.female.merge[which(male.female.merge$b.x > 0 & male.female.merge$b.y > 0),] #326 transcripts.
#172 transcripts

male.female.merge.down <- male.female.merge[which(male.female.merge$b.x < 0 & male.female.merge$b.y < 0),] #326 transcripts.
#137 transcripts

#In total, 309 of the 326 overlapping transcripts showed the same directionality of differential expression
#in both sexes.

#write csv of male/female overlapping transcripts:
write.csv(male.female.merge, "/Users/JoeKochmanski/Dropbox/Bernstein_Lab/Random TSMM Stuff/AnnaStoll_RNAseq/RNAseq_Sleuth_DGE_Male+Female_Overlap_TestTable_060921.csv")
```



```{r OPTIONAL: Pull out TPM for transcripts that show a particular direction of change}

#Pull out TPM for 172 transcripts that show downregulation:

#First read in .csv of upregulated genes:
male.female.dte.up <- read.csv("/Users/JoeKochmanski/Dropbox/Bernstein_Lab/Random TSMM Stuff/AnnaStoll_RNAseq/RNAseq_Sleuth_DGE_Male+Female_Upregulated_GeneIDs_v3_102021.csv")

#Convert sleuth to tpm table for downstream processing
#Example code:
#sleuth_to_matrix(obj, which_df, which_units)
#parameters:
#obj - a sleuth object
#which_df - character vector of length one. Which type of data to use ("obs_norm" or "obs_raw")
#which_units - character vector of length one. Which units to use ("tpm" or "est_counts")
library(tidyr)
so.F.tpm <- sleuth_to_matrix(so.F, which_df="obs_norm", which_units="tpm")
so.M.tpm <- sleuth_to_matrix(so.M, which_df="obs_norm", which_units="tpm")

colnames(so.M.tpm)
head(so.M.tpm)

#Note: We suggest removing features whose counts are consistently low (for example, removing all features that have a count of less than say 10 in more than 90% of the samples) 
#because such low-expressed features tend to reflect noise and correlations based on counts that are mostly zero aren't really meaningful. 
#The actual thresholds should be based on experimental design, sequencing depth and sample counts.

#Establish data frame of results:
so.M.tpm1 <- as.data.frame(so.M.tpm)
so.M.tpm1$target_id <- rownames(so.M.tpm1)
so.F.tpm1 <- as.data.frame(so.F.tpm)
so.F.tpm1$target_id <- rownames(so.F.tpm1)

#Now let's filter down the sleuth objects to only the consistently upregulated results.
male.female.dte.up.tpm <- left_join(male.female.dte.up, so.M.tpm1, by = "target_id") 
male.female.dte.up.tpm1 <- left_join(male.female.dte.up.tpm, so.F.tpm1, by = "target_id") 

write.csv(male.female.dte.up.tpm1, "/Users/JoeKochmanski/Dropbox/Bernstein_Lab/Random TSMM Stuff/AnnaStoll_RNAseq/RNAseq_Sleuth_DGE_Male+Female_Upregulated_TPM_110821.csv")
```

```{r OPTIONAL: Another example for pulling out TPM for transcripts of interest}

#Pull out TPM for transcripts in DropViz astrocytes+microglia that show consistent upregulation:

#First read in .csv of upregulated genes:
male.female.astromicro.dte.up <- read.csv("/Users/JoeKochmanski/Dropbox/Bernstein_Lab/Random TSMM Stuff/AnnaStoll_RNAseq/RNAseq_Sleuth_DGE_Male+Female_Overlap_TestTable_Astro+Micro_Upregulated_120921.csv")

#Convert sleuth to tpm table for downstream processing
load("/Users/JoeKochmanski/Dropbox/Bernstein_Lab/Random TSMM Stuff/AnnaStoll_RNAseq/AnnaStoll_DGE_sleuth_object_male_LRT_060921.RData")
load("/Users/JoeKochmanski/Dropbox/Bernstein_Lab/Random TSMM Stuff/AnnaStoll_RNAseq/AnnaStoll_DGE_sleuth_object_female_LRT_040921.RData")

#Example code:
#sleuth_to_matrix(obj, which_df, which_units)
#parameters:
#obj - a sleuth object
#which_df - character vector of length one. Which type of data to use ("obs_norm" or "obs_raw")
#which_units - character vector of length one. Which units to use ("tpm" or "est_counts")
library(tidyr)
library(dplyr)
library(sleuth)
so.F.tpm <- sleuth_to_matrix(so.F, which_df="obs_norm", which_units="tpm")
so.M.tpm <- sleuth_to_matrix(so.M, which_df="obs_norm", which_units="tpm")
colnames(so.M.tpm)
head(so.M.tpm)

#Note: We suggest removing features whose counts are consistently low (for example, removing all features that have a count of less than say 10 in more than 90% of the samples) 
#because such low-expressed features tend to reflect noise and correlations based on counts that are mostly zero aren't really meaningful. 
#The actual thresholds should be based on experimental design, sequencing depth and sample counts.

#Establish data frame of results:
so.M.tpm1 <- as.data.frame(so.M.tpm)
so.M.tpm1$target_id <- rownames(so.M.tpm1)
so.F.tpm1 <- as.data.frame(so.F.tpm)
so.F.tpm1$target_id <- rownames(so.F.tpm1)

#Now let's filter down the sleuth objects to only the consistently upregulated results.
male.female.dte.up.tpm <- left_join(male.female.astromicro.dte.up, so.M.tpm1, by = "target_id") 
male.female.dte.up.tpm1 <- left_join(male.female.dte.up.tpm, so.F.tpm1, by = "target_id") 

#Now let's generate averages of the male and female data by group:
head(male.female.dte.up.tpm1)
# data for rowmeans in r example
male.female.dte.up.tpm1$Male_Control_MeanTPM <- rowMeans(male.female.dte.up.tpm1[,c(47,48,54)])
male.female.dte.up.tpm1$Male_PFF_MeanTPM <- rowMeans(male.female.dte.up.tpm1[,c(49,50,51,52,53)])
male.female.dte.up.tpm1$Female_Control_MeanTPM <- rowMeans(male.female.dte.up.tpm1[,c(59,60,61,62)])
male.female.dte.up.tpm1$Female_PFF_MeanTPM <- rowMeans(male.female.dte.up.tpm1[,c(55,56,57,58)])

write.csv(male.female.dte.up.tpm1, "/Users/JoeKochmanski/Dropbox/Bernstein_Lab/Random TSMM Stuff/AnnaStoll_RNAseq/RNAseq_Sleuth_DGE_Male+Female_Overlap_TestTable_Astro+Micro_Upregulated_MeanTPM_121321.csv")

```

